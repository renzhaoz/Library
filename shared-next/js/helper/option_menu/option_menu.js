/* eslint-disable no-undef, no-invalid-this */
/*
 * Generic action menu. Options should have the following structure:
 * //The priority is the show how to order menu items.
 * //Priority "1" is the left button, "2" is the second button, rest of all items are groupped and ordered by priority in the right "Options" menu.
 *
 * new OptionMenu(options);
 *
 * options {
 *
 * items: An array of menu options to render
 * eg.
 * [
 * {
 * name: 'Lorem ipsum',
 * l10nId: 'lorem',
 * l10nArgs: 'ipsum',
 * method: function optionMethod(param1, param2) {
 * // Method and params if needed
 * },
 * params: ['param1', '123123123'],
 * priority: 1 //
 * },
 * ....
 * ,
 *
 *
 * Last option has a different UI compared with the previous one.
 * This is because it's recommended to use as a 'Cancel' option
 * {
 * name: 'Cancel',
 * l10nId: 'Cancel'
 * method: function optionMethod(param) {
 * // Method and param if needed
 * },
 * params: ['Optional params'],
 *
 * // Optional boolean flag to tell the
 * // menu button handlers that this option
 * // will not execute the "complete" callback.
 * // Defaults to "false"
 *
 * incomplete: false [true]
 * }
 * ],
 *
 * // Optional header text or node
 * header: ...,
 *
 * // additional classes on the dialog, as an array of strings
 * classes: ...
 *
 * // Optional section text or node
 * section: ...
 *
 * // Optional data-type: confirm or action
 * type: 'confirm'
 *
 * // Optional callback to be invoked when a
 * // button in the menu is pressed. Can be
 * // overridden by an "incomplete: true" set
 * // on the menu item in the items array.
 * complete: function() {...}
 * }
 */

const SUB_MENU_PREFIX = 'submenu_';
let secondCall = false;
const OptionMenu = function OptionMenu(options, optionMenuCallback) {
  if (!options || !options.items || options.items.length === 0) {
    return;
  }
  // eslint-disable-next-line no-invalid-this
  this.optionMenuCallback = optionMenuCallback;
  /*
   * Create a private, weakly held entry for
   * this instances DOM object references
   * More info:
   * https://developer.mozilla.org/en-US/docs/JavaScript/Reference/Global_Objects/WeakMap
   */
  const handlers = new WeakMap();
  // Retrieve items to be rendered
  const mainItems = [];
  const subItemsMap = {};
  let { menuClassName } = options;
  if (!menuClassName) {
    menuClassName = 'menu-button';
  }
  const self = this;
  const menuPre = document.getElementById('option-menu');
  if (menuPre && menuPre.parentNode) {
    menuPre.remove();
  }
  let header = null;
  // Create the structure
  this.form = document.createElement('form');
  this.form.id = 'option-menu';
  this.form.dataset.type = options.type || 'action';
  this.form.dataset.subtype = 'menu';
  this.form.setAttribute('role', 'dialog');
  this.form.tabIndex = -1;
  this.subMenuDisplayed = false;

  const { classList } = this.form;

  if (options.classes) {
    // eslint-disable-next-line prefer-spread
    classList.add.apply(classList, options.classes);
  }

  if (options.section) {
    const section = document.createElement('section');

    if (typeof options.section === 'string') {
      section.textContent = options.section || '';
    } else {
      section.appendChild(options.section);
    }

    this.form.appendChild(section);
  }

  getMenuItems(options.items);

  // We append a menu with the list of options
  const menu = document.createElement('menu');
  menu.id = 'mainmenu';
  menu.dataset.items = mainItems.length;

  // We append title if needed
  if (options.header) {
    header = document.createElement('h1');
    if (typeof options.header === 'string') {
      header.textContent = options.header || '';
    } else if (options.header.l10nId) {
      header.setAttribute('data-l10n-id', options.header.l10nId);
    } else {
      header.appendChild(options.header);
    }

    this.form.appendChild(header);
  }

  // eslint-disable-next-line init-declarations
  let firstItemId;
  // For each option, we append the item and listener
  mainItems.forEach(function renderOption(item) {
    const button = document.createElement('button');
    button.type = 'button';
    button.classList.add(menuClassName);
    button.classList.add('p-pri');
    if (item.id) button.id = item.id;
    if (item.l10nId) {
      window.api.l10n.setAttributes(button, item.l10nId, item.l10nArgs);
      if (button.textContent === '') {
        button.textContent = item.name || '';
      }
    } else if (item.name && item.name.length) {
      button.textContent = item.name || '';
    } else {
      // No l10n or name, just empty item, don't add to the menu
      return;
    }
    if (item.disable) {
      button.classList.add('disabled');
    }
    menu.appendChild(button);
    if (!firstItemId) {
      if (!button.id) {
        button.id = 'option-menu-first-item';
      }
      firstItemId = button.id;
    }
    /*
     * Add a mapping from the button object
     * directly to its options item.
     */
    item.incomplete = item.incomplete || false;

    handlers.set(button, item);
  });

  this.form.addEventListener('submit', function submitFunc(event) {
    event.preventDefault();
  });

  this.form.addEventListener(
    'transitionend',
    function transitionend(event) {
      if (event.target !== this.form) {
        return;
      }

      if (!this.form.classList.contains('visible')) {
        if (this.form.parentNode) {
          this.form.remove();
        }
      } else {
        this.form.focus();
      }

      document.body.classList.remove('dialog-animating');
      if (window.NavigationMap) {
        NavigationMap.lockNavigation = false;
      }
    }.bind(this)
  );

  menu.addEventListener('click', menuHandler);

  // Appending the action menu to the form
  this.form.appendChild(menu);
  if (header) {
    header.setAttribute('role', 'heading');
    header.setAttribute('aria-labelledby', firstItemId);
  }
  genSubMenus.call(this);

  function menuHandler(event) {
    if (secondCall) return;
    secondCall = true;

    const { target } = event;
    if ('true' === target.dataset.hasmenu) {
      self.subMenuDisplayed = true;
      menuOwnerHandler(target);
      secondCall = false;
      return;
    }

    let action = handlers.get(target);
    // eslint-disable-next-line init-declarations
    let method;
    if (!action && target.tagName.toLowerCase() === 'menu') {
      const focusItem = target.querySelector('.focus');
      if (focusItem) {
        focusItem.focus();
        action = handlers.get(focusItem);
      }
    }
    /*
     * Delegate operation to target method. This allows
     * for a custom "Cancel" to be provided by calling program.
     *
     * Further operations should only be processed if
     * an actual button was pressed.
     */
    if (typeof action !== 'undefined') {
      // eslint-disable-next-line no-empty-function
      method = action.method || function func() {};

      // eslint-disable-next-line prefer-spread
      method.apply(null, action.params || []);

      // Hide action menu when click is received
      self.hide();

      if (typeof options.complete === 'function' && !action.incomplete) {
        options.complete();
      }
    }
  }

  function menuOwnerHandler(owner) {
    const subid = SUB_MENU_PREFIX + owner.id;
    const submenu = getSubMenu(subid);
    if (submenu) {
      // Submenu.style.bottom = Math.max(window.innerHeight - target.offsetTop - target.offsetHeight - submenu.offsetHeight, 0) + 'px';
      setTimeout(() => {
        self.toggleMainMenu(false);
        menu.classList.add('submenu-displayed');
        owner.classList.add('submenu-owner');
        submenu.classList.remove('hidden');
        self._setFocus(submenu.children[1]);
      }, 400);
    }

    function getSubMenu(id) {
      for (let i = 1; i < self.form.children.length; i++) {
        const item = self.form.children[i];
        if (item.id === id) return item;
      }
      return null;
    }
  }

  // Used for getting the items of both mainmenu and submenu
  function getMenuItems(items) {
    for (const item of items) {
      const { fid } = item;
      if (fid) {
        if (subItemsMap[fid]) subItemsMap[fid].push(item);
        else subItemsMap[fid] = [item];
      } else {
        mainItems.push(item);
      }
    }
  }

  // Used for generating submenu
  function genSubMenus() {
    for (const fid in subItemsMap) {
      const subItems = subItemsMap[fid];
      // eslint-disable-next-line no-continue
      if (!Array.isArray(subItems) || subItems.length === 0) continue;
      console.log('genSubMenus', fid, subItems);
      const submenu = document.createElement('menu');
      submenu.id = SUB_MENU_PREFIX + fid;
      submenu.setAttribute('fid', fid);
      submenu.dataset.subtype = 'submenu';
      submenu.dataset.items = subItems.length;
      submenu.classList.add('hidden');
      submenu.setAttribute('role', 'heading');
      submenu.setAttribute('aria-labelledby', 'sub-menu-header');
      genSubHeader(submenu, fid);
      for (const item of subItems) {
        const button = document.createElement('button');
        button.type = 'button';
        button.classList.add(menuClassName);
        button.classList.add('p-pri');
        if (item.id) button.id = item.id;
        if (item.l10nId)
          window.api.l10n.setAttributes(button, item.l10nId, item.l10nArgs);
        else button.textContent = item.name;
        item.incomplete = item.incomplete || false;
        handlers.set(button, item);
        submenu.appendChild(button);
      } // For
      submenu.addEventListener(
        'click',
        function click(event) {
          this.subMenuDisplayed = false;
          menuHandler(event);
          // eslint-disable-next-line no-invalid-this
        }.bind(this)
      );
      this.form[fid].dataset.hasmenu = 'true';
      this.form.appendChild(submenu);
    } // For

    function genSubHeader(submenu, fid) {
      const item = mainItems.find(i => i.id === fid);
      if (item) {
        const subHeader = document.createElement('header');
        subHeader.id = 'sub-menu-header';
        subHeader.setAttribute('role', 'heading');
        if (item.l10nId) {
          window.api.l10n.setAttributes(subHeader, item.l10nId, item.l10nArgs);
        } else {
          subHeader.textContent = item.name;
        }
        submenu.appendChild(subHeader);
      }
    }
  } // GenSubMenus
};

OptionMenu.prototype._setFocus = function _setFocus(element) {
  const focused = document.querySelector('.focus');
  if (focused) focused.classList.remove('focus');
  element.setAttribute('tabindex', 1);
  element.classList.add('focus');
  element.focus();
};

// We prototype functions to show/hide the UI of action-menu
OptionMenu.prototype.show = function show() {
  if (window.NavigationMap) {
    NavigationMap.lockNavigation = true;
  }
  // Remove the focus to hide the keyboard asap
  // eslint-disable-next-line no-unused-expressions
  document.activeElement && document.activeElement.blur();

  if (!this.form.parentNode) {
    document.body.appendChild(this.form);

    /*
     * Flush style on form so that the show transition plays once we add
     * the visible class.
     */
    // eslint-disable-next-line no-unused-expressions
    this.form.clientTop;
  }
  this.form.classList.add('visible');
  document.body.classList.add('dialog-animating', 'show-option');
  if (this.optionMenuCallback) {
    this.optionMenuCallback(true);
  }

  const menu = document.getElementById('mainmenu');
  const header = this.form.querySelector('h1');
  if (menu && header) {
    // The setTimeout can to avoid getting menu.offsetHeight when pop animation is showing
    setTimeout(() => {
      header.setAttribute('style', `bottom:${menu.offsetHeight}px`);
    });
  }

  return Promise.resolve();
};

OptionMenu.prototype.toggleMainMenu = function toggleMainMenu(display) {
  const header = this.form.querySelector('h1');
  const mainmenu = this.form.querySelector('#mainmenu');
  if (display) {
    mainmenu.classList.remove('hidden');
    header.classList.remove('hidden');
  } else {
    mainmenu.classList.add('hidden');
    header.classList.add('hidden');
  }
};

OptionMenu.prototype.hide = function hide() {
  if (!this.form.classList.contains('visible')) {
    return;
  }

  if (this.subMenuDisplayed) {
    const selector = 'form[data-subtype=menu] > menu > button.submenu-owner';
    const owner = document.querySelector(selector);
    if (owner) {
      const submenu = document.getElementById(SUB_MENU_PREFIX + owner.id);
      if (submenu) {
        submenu.classList.add('hidden');
        owner.classList.remove('submenu-owner');
        owner.parentElement.classList.remove('submenu-displayed');
        this.subMenuDisplayed = false;
        this.toggleMainMenu(true);
        this._setFocus(owner);
        secondCall = false;
        // eslint-disable-next-line consistent-return
        return new Promise(resolve => {
          resolve();
        });
      }
    }
  }
  this.form.setAttribute('aria-hidden', 'true');
  this.form.classList.remove('visible');
  if (this.optionMenuCallback) {
    this.optionMenuCallback(false);
  }

  setTimeout(() => {
    document.body.classList.remove('show-option');
  }, 10);

  // eslint-disable-next-line consistent-return
  return new Promise(resolve => {
    const timer = setTimeout(() => {
      resolve();
    }, 300);

    this.form.addEventListener('transitionend', () => {
      if (timer) {
        clearTimeout(timer);
      }
      resolve();
    });
  });
};

window.OptionMenu = OptionMenu;
