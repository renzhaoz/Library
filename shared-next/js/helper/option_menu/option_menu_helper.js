/* global SoftkeyPanel*/
/* eslint-disable no-undef, no-unused-expressions */

(function optionMenuHelper(exports) {
  const buttonState = {
      leftKeyState: '',
      _cacheParams: null
    },
    firtsUpper = function firtsUpper(str) {
      return str.charAt(0).toUpperCase() + str.slice(1);
    };
  const OptionHelper = function OptionHelper() {
    this.option = {};
    this.optionParams = Object.create(null);
    this.softkeyPanel = undefined;
    this.optionsCallback = undefined;
    this.lastParamName = undefined;
  };
  OptionHelper.prototype.stopKeyEvents = function stopKeyEvents() {
    this.softkeyPanel.stopListener();
  };
  OptionHelper.prototype.startKeyEvents = function startKeyEvents() {
    this.softkeyPanel.startListener();
  };
  OptionHelper.prototype.setOptionMenuCallBack = function setOptionMenuCallBack(
    callback
  ) {
    this.optionsCallback = callback;
  };
  OptionHelper.prototype._initSoftKeyPanel = function _initSoftKeyPanel() {
    if (!this.softkeyPanel) {
      console.log('Option helper, _initSoftKeyPanel');
      if (typeof SoftkeyHelper !== 'undefined') {
        // If we have SoftkeyHelper we don't need to implement new softkeypanel
        SoftkeyHelper.init(this._params, this.optionsCallback);
        this.softkeyPanel = SoftkeyHelper.getSoftkey();
      } else {
        this.softkeyPanel = new SoftkeyPanel(
          this._params,
          this.optionsCallback
        );
      }
      this.softkeyPanel.show();
      return;
    }
    this.softkeyPanel.initSoftKeyPanel(this._params);
  };
  OptionHelper.prototype.show = function show(paramName) {
    this.lastParamName = paramName;
    console.log(`Option helper, show paramName: ${paramName}`);
    if (this.softkeyPanel && !this.softkeyPanel.softKeyVisible) {
      this.softkeyPanel.show();
    }
    try {
      if (this.optionParams[paramName]) {
        this._params = this.getValidParams(paramName);
        window.api.l10n.once(this._initSoftKeyPanel.bind(this));
      } else {
        console.log(`Param <${paramName}> not found`);
      }
    } catch (e) {
      console.log(e);
    }
  };
  OptionHelper.prototype.getValidParams = function getValidParams(paramName) {
    const params = this.optionParams[paramName];
    if (typeof params.items === 'undefined') {
      return params;
    }

    const validParams = {};
    const validItems = [];
    for (const p in params) {
      validParams[p] = params[p];
    }
    for (let i = 0; i < validParams.items.length; i++) {
      const item = validParams.items[i];
      if (item.l10nId || (item.name && item.name.length) || item.icon) {
        validItems.push(item);
      }
    }
    validParams.items = validItems;
    return validParams;
  };
  OptionHelper.prototype.popAndRemoveElement = function popAndRemoveElement(
    arr,
    paramName
  ) {
    // eslint-disable-next-line init-declarations
    let item;
    arr.forEach(params => {
      if (params.name === paramName) {
        const index = arr.indexOf(params);
        if (index >= 0) {
          item = arr.splice(index, 1);
        }
      }
    });
    /*
     * If (item == undefined) {
     * return item;
     *}
     */
    return (item && item[0]) || undefined;
  };
  OptionHelper.prototype.getParams = function getParams(paramName) {
    return this.optionParams[paramName];
  };
  OptionHelper.prototype.getLastParamName = function getLastParamName() {
    return this.lastParamName;
  };
  OptionHelper.prototype.compare = function compare(a, b) {
    return a.priority - b.priority;
  };

  // eslint-disable-next-line max-params
  OptionHelper.prototype.swapParams = function swapParams(
    paramsName,
    oldName,
    newName,
    render
  ) {
    const thisParms = this.getParams(paramsName);
    const itemToHide = this.popAndRemoveElement(thisParms.items, oldName);
    if (itemToHide) {
      thisParms.hidenItems.push(itemToHide);
      thisParms.items.push(
        this.popAndRemoveElement(thisParms.hidenItems, newName)
      );
      thisParms.items.sort(this.compare);
      if (render) {
        this.show(paramsName);
      }
    }
  };

  OptionHelper.prototype.showParams = function showParams(
    paramsName,
    name,
    render
  ) {
    const thisParms = this.getParams(paramsName);
    const itemToShow = this.popAndRemoveElement(thisParms.hidenItems, name);
    if (itemToShow) {
      thisParms.items.push(itemToShow);
    }
    thisParms.items.sort(this.compare);
    if (render) {
      this.show(paramsName);
    }
  };
  OptionHelper.prototype.hideParams = function hideParams(
    paramsName,
    name,
    render
  ) {
    const thisParms = this.getParams(paramsName);
    const itemToHide = this.popAndRemoveElement(thisParms.items, name);
    // Condition added to prevent adding unncessary "undefined" into array
    if (itemToHide) {
      thisParms.hidenItems.push(itemToHide);
    }
    thisParms.items.sort(this.compare);
    if (render) {
      this.show(paramsName);
    }
  };
  OptionHelper.prototype.hideMenu = function hideMenu() {
    this.softkeyPanel && this.softkeyPanel.hide();
    this.softkeyPanel.stopListener();
  };
  OptionHelper.prototype.showMenu = function showMenu() {
    this.softkeyPanel && this.softkeyPanel.show();
    this.softkeyPanel.startListener();
  };
  OptionHelper.prototype.saveContext = function saveContext() {
    if (this.softkeyPanel) {
      const leftKey = this.softkeyPanel.getLeftKey();
      buttonState.leftKeyState = leftKey.classList.contains('hide');
      leftKey.classList.remove('hide');
      buttonState._cacheParams = this.getLastParamName();
    }
  };
  OptionHelper.prototype.returnContext = function returnContext() {
    if (this.softkeyPanel) {
      this.show(buttonState._cacheParams);
      buttonState.leftKeyState &&
        this.softkeyPanel.getLeftKey().classList.add('hide');
    }
  };
  OptionHelper.prototype.getLeftKey = function getLeftKey() {
    return this.softkeyPanel.getLeftKey();
  };
  OptionHelper.prototype.setLast = function setLast(selector) {
    buttonState._cacheParams = selector;
  };
  OptionHelper.prototype.getSoftKeyByName = function getSoftKeyByName(key) {
    // eslint-disable-next-line init-declarations
    let button;
    try {
      button = this.softkeyPanel[`get${firtsUpper(key)}Key`]();
    } catch (err) {
      // If some st dev try to get not being existed button
      console.error(err);
      return null;
    }
    return button;
  };
  /*
   * Get softkey button [left, center, right] and if it exists check visibility state if it does not hide return true
   * @param String one of [left, center, right] default value is left
   * @return Boolean
   */
  OptionHelper.prototype.isAvailable = function isAvailable(key) {
    let result = false;
    const button = this.getSoftKeyByName(key || (key = 'left'));
    button && (result = !button.classList.contains('hide'));
    return result;
  };
  /*
   * Do softkey btns visible or invisible
   * @param btn String one of [left, center, right] default value is left
   * @param state String one of [show, hide] default value is show
   * @void
   */
  OptionHelper.prototype.changeBtnState = function changeBtnState(btn, state) {
    const button = this.getSoftKeyByName(btn || (btn = 'left'));
    try {
      button && this[`do${firtsUpper(state || (state = 'show'))}Btn`](button);
    } catch (err) {
      // If the same as first dev try to change not existed state
      console.error(err);
    }
  };

  OptionHelper.prototype.doShowBtn = function doShowBtn(button) {
    button.classList.remove('hide');
  };

  OptionHelper.prototype.doHideBtn = function doHideBtn(button) {
    button.classList.add('hide');
  };

  exports.OptionHelper = new OptionHelper();
})(window);
