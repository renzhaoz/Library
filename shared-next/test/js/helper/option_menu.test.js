/* eslint-disable no-undef, global-require */
describe('helper js <option_menu> test', () => {
  const { wrap } = require('jest-snapshot-serializer-raw');
  let optionsMenu = null;
  const callback = jest.fn();

  beforeAll(done => {
    require('../../mock/l10n_mock');
    require('../../mock/navigation_map');
    require('../../../js/helper/option_menu/option_menu');
    done();
  });

  // Test OptionMenu
  test('OptionMenu should be function', done => {
    expect(typeof window.OptionMenu).toBe('function');
    done();
  });

  // Test OptionMenu
  test('OptionMenu should be not render', done => {
    optionsMenu = new OptionMenu({});
    expect(typeof optionsMenu).toBe('object');
    done();
  });

  describe('OptionMenu function test', () => {
    let optionsMenu = null;
    beforeEach(done => {
      const options = {
        items: [
          {
            name: 'Lorem ipsum',
            l10nId: 'lorem',
            l10nArgs: 'ipsum',
            method: function method() {
              // Method and params if needed
            },
            params: ['param1', '123123123'],
            priority: 4,
            disable: true
          }
        ],
        section: 'section',
        header: 'header',
        classes: ['containerClass']
      };
      optionsMenu = new OptionMenu(options, callback);
      jest.useFakeTimers();
      optionsMenu.show();
      jest.runAllTimers();
      done();
    });

    afterEach(done => {
      jest.resetAllMocks();
      done();
    });

    // Test OptionMenu
    test('OptionMenu should be renders correctly', done => {
      expect(wrap(optionsMenu.form)).toMatchSnapshot();
      done();
    });

    // Test OptionMenu.show()
    test('OptionMenu should be shows correctly', done => {
      expect(NavigationMap.lockNavigation).toBeTruthy();
      expect(optionsMenu.form.classList.contains('visible')).toBeTruthy();
      expect(document.body.classList.contains('dialog-animating')).toBeTruthy();
      expect(document.body.classList.contains('show-option')).toBeTruthy();
      expect(setTimeout.mock.calls.length).toBe(1);
      expect(document.querySelector('h1').hasAttribute('style')).toBeTruthy();
      expect(callback.mock.calls[0][0]).toBe(true);
      done();
    });

    // Test click eventListener on menu
    test('click eventListener on menu be invokes correctly', done => {
      const clickEvent = new window.CustomEvent('click');
      document.getElementById('mainmenu').dispatchEvent(clickEvent);
      done();
    });

    // Test submit eventListener
    test('submit eventListener be invokes correctly', done => {
      const submitEvent = new window.CustomEvent('submit');
      optionsMenu.form.dispatchEvent(submitEvent);
      done();
    });

    // Test transitionend eventListener
    test('transitionend eventListener be invokes correctly', done => {
      const transitionendEvent = new window.CustomEvent('transitionend');
      optionsMenu.form.dispatchEvent(transitionendEvent);
      expect(NavigationMap.lockNavigation).toBeFalsy();
      expect(document.body.classList.contains('dialog-animating')).toBeFalsy();
      done();
    });

    // Test OptionMenu._setFocus()
    test('_setFocus be invokes correctly', done => {
      optionsMenu._setFocus(optionsMenu.form);
      expect(optionsMenu.form.hasAttribute('tabindex')).toBeTruthy();
      expect(optionsMenu.form.classList.contains('focus')).toBeTruthy();
      done();
    });

    // Test OptionMenu.toggleMainMenu()
    test('toggleMainMenu be invokes correctly', done => {
      const header = document.querySelector('h1');
      const mainmenu = document.querySelector('#mainmenu');
      optionsMenu.toggleMainMenu(true);
      expect(header.classList.contains('hidden')).toBeFalsy();
      expect(mainmenu.classList.contains('hidden')).toBeFalsy();
      optionsMenu.toggleMainMenu(false);
      expect(header.classList.contains('hidden')).toBeTruthy();
      expect(mainmenu.classList.contains('hidden')).toBeTruthy();
      done();
    });

    // Test OptionMenu.hide()
    test('hide be invokes correctly', done => {
      jest.useFakeTimers();
      optionsMenu.hide();
      expect(optionsMenu.form.hasAttribute('aria-hidden')).toBeTruthy();
      expect(optionsMenu.form.classList.contains('visible')).toBeFalsy();
      expect(callback.mock.calls[0][0]).toBe(true);
      expect(callback.mock.calls[1][0]).toBe(false);
      jest.runAllTimers();
      expect(setTimeout.mock.calls.length).toBe(2);
      expect(document.body.classList.contains('show-option')).toBeFalsy();
      const transitionendEvent = new window.CustomEvent('transitionend');
      optionsMenu.form.dispatchEvent(transitionendEvent);
      expect(clearTimeout.mock.calls.length).toBe(1);
      done();
    });
  });
});
