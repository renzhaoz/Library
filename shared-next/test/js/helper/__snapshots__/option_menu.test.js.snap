// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`helper js <option_menu> test OptionMenu function test OptionMenu should be renders correctly 1`] = `
Object {
  Symbol(jest-snapshot-serializer-raw): <form
    class="containerClass visible"
    data-subtype="menu"
    data-type="action"
    id="option-menu"
    role="dialog"
    tabindex="-1"
  >
    <section>
      section
    </section>
    <h1
      aria-labelledby="option-menu-first-item"
      role="heading"
      style="bottom:0px"
    >
      header
    </h1>
    <menu
      data-items="1"
      id="mainmenu"
    >
      <button
        class="menu-button p-pri disabled"
        data-l10n-args="\\"ipsum\\""
        data-l10n-id="lorem"
        id="option-menu-first-item"
        type="button"
      >
        Lorem ipsum
      </button>
    </menu>
  </form>,
}
`;
